lib/web_ui/lib/src/engine/frame_reference.dart

```diff
@@ -2,28 +2,61 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-/// A monotonically increasing frame number being rendered.
+/// Holds a resource that's valid until the end of the current frame.
 ///
-/// Used for debugging only.
-int debugFrameNumber = 1;
+/// If a frame-scoped object is created outside a frame, it is disposed of at
+/// the end of the next frame.
+///
+/// At the end of the frame the [disposeFrameResources] method is called automatically.
+abstract class FrameScoped {
+  FrameScoped() {
+    _scopedObjects.add(this);
```

Comment:
I'm not sure this pattern is actually going to work as intended here. Because we are adding to `_scopedObjects` only on construction, `disposeFrameResources` will only ever be called once in the lifetime of the object. Which means that this won't do the thing you want it to do if the `CkPaint` object is reused across frames (you also want to dispose things after the reuse as well).

The way I'm thinking about this problem is that the `CkPaint` object contains all the dart-side state that is reified into the Skia-side object whenever needed. The `CkPaint` object may exist across multiple frames. So I think the `CkPaint` needs to have (as one of its fields) a container object that wraps the `SkPaint` and that container object is managed by this scoping system. I don't think `CkPaint` should directly extend this class.

Reply:
Actually, `FrameScoped` is intentionally designed to be reusable across frames. When `disposeFrameResources` is called it does not completely destroy the `FrameScoped` object. All it does is release any frame resources that the object may have accumulated. It's totally fine to reuse the object on the next frame, and if that happens the object may accumulate new frame-specific resources for that frame too.

It may be useful to think about `FrameScoped` and "frame resource` as two separate entities, so when `disposeFrameResources` is called you only dispose the "frame resource" part, but not the `FrameScoped` part:

```mermaid
erDiagram
    FRAME-SCOPED ||..|{ FRAME-RESOURCE : has
    FRAME-RESOURCE
```

Maybe I need a better name than `FrameScoped` because it implies that the object itself is scoped to one frame and not reusable. Maybe something like `HasFrameResources`?

Reply:
Maybe this is too generic but perhaps this concept could be generalized to be "objects which want to be informed that the frame has ended" (`FrameAware`?) For `CkPaint`, it wants to know the framed ended so it can dispose of its `SkPaint` object, and `FrameScopedValue` would also be a valid and useful type of `FrameAware` object. But the concept wouldn't be limited to just objects which want to dispose of resources on frame end.

Reply:
Actually `FrameAware` is brilliant! I think the only level of specificity we care about is "frame". Everything else can be infinitely generic. It shouldn't matter what you need to do at the end of the frame, dispose of resources or perform other end-of-frame logic.

This also made me realize that I have a bug in my code. After `FrameScoped` is notified about the end of the frame, it is never notified again. It needs to re-register every time it is populated with frame resources again. I'll have to fix that.